### PR TITLE
Replaces Webpack with Esbuild for faster bundling

### PR DESCRIPTION
**Pull request by heyjordn**
Hi, great work on making this, however when I try to bundle the project, the build times are so long, I switched out Webpack with [Esbuild](https://esbuild.github.io/) and the build times are down to seconds now. 

Could you review and let me know if this could work?

### CODE REVIEW

**Comment by jjppof** [comment]
Hi! Thanks for the PR!
It seems that esbuild is really faster, but we need to have the same dev environment as webpack in order to start using it. For instance, in the way you set `npm start`, the application is just watching the changes, but not serving the application. Please check [webpack.config.js](https://github.com/jjppof/goldensun_html5/blob/master/webpack.config.js) and set esbuild to have the same behavior as webpack.
Also, don't forget to add esbuild to dev dependencies in package.json and remove webpack.

**Comment by heyjordn** [comment]
Hey @jjppof I've added in `esbuild-serve` for serving the assets along with a quick build time, testing on my end and it seems to be working

File: package.json [hunk 1]
```diff
@@ -2,14 +2,15 @@
...
   "scripts": {
-    "start": "webpack-dev-server",
-    "build": "webpack",
+    "start": "node esbuild.config.js -w",
+    "build": "node esbuild.config.js",
     "format-check": "prettier --check \"base/**/*.ts\"",
     "format": "npx prettier --write \"base/**/*.ts\"",
     "docs": "node scripts/generate_docs.js",
-    "postinstall": "husky install",
+    "_postinstall": "husky install",
```
Comment (jjppof) [inline package.json:13]: why the underscore before `postinstall`?

File: package.json [hunk 2]
```diff
@@ -26,19 +27,19 @@
...
-    "typescript": "^4.0.3",
-    "webpack": "^4.44.2",
-    "webpack-cli": "^3.3.12",
-    "webpack-dev-server": "^3.11.0"
+    "typescript": "^4.3.4"
   },
   "dependencies": {
     "lodash": "^4.17.21",
     "mathjs": "^9.3.2",
+    "phaser-ce": "^2.18.0",
```
Comment (jjppof) [inline package.json:42]: Please remove this, gshtml5 uses a custom Phaser.

File: esbuild.config.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+import esbuildServe from "esbuild-serve";
+
+esbuildServe(
+  {
+    entryPoints: ["./base/GoldenSun.ts"],
+    bundle: true,
+    minify: true,
+    sourcemap: "inline",
+    target: ["es2019", "chrome58", "firefox57"],
```
Comment (jjppof) [inline esbuild.config.js:9]: I don't think we need `chrome58` and `firefox57` targets. Is there a reason for them?

File: webpack.config.js [hunk 4]
```diff
@@ -1,39 +0,0 @@
...
-  devtool: 'inline-source-map',
-  devServer: {
-    contentBase: __dirname,
-    watchContentBase: true,
-    compress: true,
-    liveReload: true,
-    open: true,
-    port: 9000,
-    watchOptions: {
-      ignored: [
```
Comment (jjppof) [inline webpack.config.js:14]: We also need this ignore list on `esbuild-serve`.

**Review by jjppof — CHANGES_REQUESTED** [review]
Please set to `es2017` and add an ignore list in the watcher.

File: tsconfig.json [hunk 5]
```diff
@@ -17,6 +17,5 @@
         "./code_docs",
         "./scripts",
         "./static/*.js",
```
Comment (jjppof) [inline tsconfig.json:19]: Need to add esbuild.config.js to this list.

File: package.json [hunk 6]
```diff
@@ -26,15 +27,14 @@
   "devDependencies": {
     "@types/lodash": "^4.14.161",
     "@types/mathjs": "^6.0.11",
+    "esbuild": "^0.12.12",
+    "esbuild-serve": "^1.0.1",
```
Comment (jjppof) [inline package.json:31]: I was checking this package and it has only 13 stars on Github, I usually avoid packages with little engagement like this one. But it seems that esbuild already has an internal functionality to serve applications, [see](https://esbuild.github.io/api/#serve). If it fits for our case, using only esbuild is preferable.

**Review by jjppof — CHANGES_REQUESTED** [review]
Changes in /assets folder are not causing the browser to reload. Also, would be nice if on `npm start` the browser open just like we had with webpack.

**Comment by heyjordn** [comment]
Hey @jjppof I had to switch to [estrella](https://github.com/rsms/estrella), another esbuild build too to rebuild when `assets` changed, looking into the open broswer feature,

File: package.json [hunk 7]
```diff
@@ -26,15 +26,13 @@
...
     "pinst": "^2.0.0",
     "prettier": "2.1.2",
-    "ts-loader": "^8.0.4",
+    "serve-http": "^1.0.6",
     "typedoc": "^0.20.35",
-    "typescript": "^4.0.3",
-    "webpack": "^4.44.2",
-    "webpack-cli": "^3.3.12",
-    "webpack-dev-server": "^3.11.0"
+    "typescript": "^4.3.4"
```
Comment (jjppof) [inline package.json:35]: Please downgrade typescript version to 4.2.X in order to have compatibility with typedocs.

File: scripts/build.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+  bundle: true,
+  tsc: true,
+  external: ["dist/*", ".git", "code_docs", "scripts"],
+});
+
+// Watch our asset files for changes
+const dir = "assets";
+const filter = /\..*$/i;
+
+scandir(dir, filter, { recursive: true }).then((files) => {
```
Comment (jjppof) [inline scripts/build.js:20]: `files` is not being used here.

File: scripts/build.js [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+  cliopts.watch &&
+    watch(dir, { filter, recursive: true }, (changes) => {
+      changes.map((c) => console.log(`Reloading ${file} -> ${file}`));
```
Comment (jjppof) [inline scripts/build.js:23]: `c` is not being used here.

**Review by jjppof — CHANGES_REQUESTED** [review]
I got this error when changing something inside assets folder (assets\filters\reveal.js specifically):

> Unhandled exception: ReferenceError: file is not defined

Also, please change `serve-http`  (which is 24 stars only) to [`http-server`](https://github.com/http-party/http-server). Use the `-o` and `-c-1` (very important) options.

Since you are using a third-party package to serve, why not using raw `esbuild`? It also has [watch callback](https://esbuild.github.io/api/#watch).

**Comment by MilesBHuff** [comment]
@heyjordn This PR needs to merge the latest changes from `master` -- the configs are rather different now.

**Comment by jjppof** [comment]
I think we can close this one for now. We can think of a faster bundler for a next release.

**Pull request closed without merging**